File: web-ui/src/api/missions.ts
import { apiFetch, apiPost, apiDelete, apiPatch } from "./client";

// ── Mission types (v2: goal-driven loop) ──────────────

export type MissionState =
  | "pending"
  | "executing"
  | "evaluating"
  | "paused"
  | "completed"
  | "cancelled"
  | "failed";

export type EvalVerdict = "achieved" | "continue" | "blocked" | "failed";

export interface EvalRecord {
  iteration: number;
  verdict: EvalVerdict;
  summary: string;
  next_step?: string | null;
  timestamp: string;
}

export interface Mission {
  id: string;
  goal: string;
  session_id: string;
  project_index: number;
  state: MissionState;
  iteration: number;
  max_iterations: number;
  last_verdict?: EvalVerdict | null;
  last_eval_summary?: string | null;
  eval_history?: EvalRecord[];
  created_at: string;
  updated_at: string;
}

export interface MissionsListResponse {
  missions: Mission[];
}

export interface CreateMissionRequest {
  goal: string;
  session_id?: string | null;
  project_index?: number;
  max_iterations?: number;
}

export interface UpdateMissionRequest {
  goal?: string;
  max_iterations?: number;
}

export type MissionAction = "start" | "pause" | "resume" | "cancel";

// ── Mission API ───────────────────────────────────────

export async function fetchMissions(): Promise<MissionsListResponse> {
  return apiFetch<MissionsListResponse>("/missions");
}

export async function fetchMission(missionId: string): Promise<Mission> {
  return apiFetch<Mission>(`/missions/${encodeURIComponent(missionId)}`);
}

export async function createMission(req: CreateMissionRequest): Promise<Mission> {
  return apiPost<Mission>("/missions", req);
}

export async function updateMission(
  missionId: string,
  req: UpdateMissionRequest
): Promise<Mission> {
  return apiPatch<Mission>(`/missions/${encodeURIComponent(missionId)}`, req);
}

export async function deleteMission(missionId: string): Promise<void> {
  return apiDelete(`/missions/${encodeURIComponent(missionId)}`);
}

export async function missionAction(
  missionId: string,
  action: MissionAction
): Promise<Mission> {
  return apiPost<Mission>(`/missions/${encodeURIComponent(missionId)}/action`, { action });
}

// ── Personal Memory types ─────────────────────────────

export type MemoryScope = "global" | "project" | "session";

export interface PersonalMemoryItem {
  id: string;
  label: string;
  content: string;
  scope: MemoryScope;
  project_index: number | null;
  session_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface PersonalMemoryListResponse {
  memory: PersonalMemoryItem[];
}

export interface CreatePersonalMemoryRequest {
  label: string;
  content: string;
  scope: MemoryScope;
  project_index?: number | null;
  session_id?: string | null;
}

export interface UpdatePersonalMemoryRequest {
  label?: string;
  content?: string;
  scope?: MemoryScope;
  project_index?: number | null;
  session_id?: string | null;
}

// ── Personal Memory API ───────────────────────────────

export async function fetchPersonalMemory(): Promise<PersonalMemoryListResponse> {
  return apiFetch<PersonalMemoryListResponse>("/memory");
}

export async function createPersonalMemory(
  req: CreatePersonalMemoryRequest
): Promise<PersonalMemoryItem> {
  return apiPost<PersonalMemoryItem>("/memory", req);
}

export async function updatePersonalMemory(
  memoryId: string,
  req: UpdatePersonalMemoryRequest
): Promise<PersonalMemoryItem> {
  return apiPatch<PersonalMemoryItem>(`/memory/${encodeURIComponent(memoryId)}`, req);
}

export async function deletePersonalMemory(memoryId: string): Promise<void> {
  return apiDelete(`/memory/${encodeURIComponent(memoryId)}`);
}

// ── Autonomy Controls ─────────────────────────────────

export type AutonomyMode = "observe" | "nudge" | "continue" | "autonomous";

export interface AutonomySettings {
  mode: AutonomyMode;
  updated_at: string;
}

export async function fetchAutonomySettings(): Promise<AutonomySettings> {
  return apiFetch<AutonomySettings>("/autonomy");
}

export async function updateAutonomySettings(mode: AutonomyMode): Promise<AutonomySettings> {
  return apiPost<AutonomySettings>("/autonomy", { mode });
}

File: leptos-ui/src/api/missions.rs
//! Missions & Memory & Autonomy API — matches React `api/missions.ts`.

use serde::Serialize;
use crate::types::api::{
    Mission, MissionsListResponse, PersonalMemoryItem, PersonalMemoryListResponse,
    AutonomySettings,
};
use super::client::{api_fetch, api_post, api_delete, api_patch, ApiError};

// ── Request types ───────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct CreateMissionRequest {
    pub goal: String,
    pub session_id: String,
    pub project_index: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_iterations: Option<u32>,
}

#[derive(Debug, Clone, Serialize)]
pub struct UpdateMissionRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub goal: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_iterations: Option<u32>,
}

#[derive(Debug, Clone, Serialize)]
pub struct CreateMemoryRequest {
    pub label: String,
    pub content: String,
    pub scope: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_index: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct UpdateMemoryRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scope: Option<String>,
}

// ── Missions API ────────────────────────────────────────────────────

/// List all missions.
pub async fn fetch_missions() -> Result<Vec<Mission>, ApiError> {
    let resp: MissionsListResponse = api_fetch("/missions").await?;
    Ok(resp.missions)
}

/// Get a single mission.
pub async fn fetch_mission(mission_id: &str) -> Result<Mission, ApiError> {
    let path = format!(
        "/missions/{}",
        js_sys::encode_uri_component(mission_id),
    );
    api_fetch(&path).await
}

/// Create a new mission.
pub async fn create_mission(req: &CreateMissionRequest) -> Result<Mission, ApiError> {
    api_post("/missions", req).await
}

/// Update a mission.
pub async fn update_mission(mission_id: &str, req: &UpdateMissionRequest) -> Result<Mission, ApiError> {
    let path = format!(
        "/missions/{}",
        js_sys::encode_uri_component(mission_id),
    );
    api_patch(&path, req).await
}

/// Delete a mission.
pub async fn delete_mission(mission_id: &str) -> Result<(), ApiError> {
    let path = format!(
        "/missions/{}",
        js_sys::encode_uri_component(mission_id),
    );
    api_delete(&path).await
}

/// Perform a mission action (start, pause, resume, cancel).
pub async fn mission_action(mission_id: &str, action: &str) -> Result<Mission, ApiError> {
    #[derive(Serialize)]
    struct Body<'a> {
        action: &'a str,
    }
    let path = format!(
        "/missions/{}/action",
        js_sys::encode_uri_component(mission_id),
    );
    api_post(&path, &Body { action }).await
}

// ── Personal Memory API ─────────────────────────────────────────────

/// List personal memory items.
pub async fn fetch_personal_memory() -> Result<Vec<PersonalMemoryItem>, ApiError> {
    let resp: PersonalMemoryListResponse = api_fetch("/memory").await?;
    Ok(resp.memory)
}

/// Create a personal memory item.
pub async fn create_personal_memory(req: &CreateMemoryRequest) -> Result<PersonalMemoryItem, ApiError> {
    api_post("/memory", req).await
}

/// Update a personal memory item.
pub async fn update_personal_memory(id: &str, req: &UpdateMemoryRequest) -> Result<PersonalMemoryItem, ApiError> {
    let path = format!(
        "/memory/{}",
        js_sys::encode_uri_component(id),
    );
    api_patch(&path, req).await
}

/// Delete a personal memory item.
pub async fn delete_personal_memory(id: &str) -> Result<(), ApiError> {
    let path = format!(
        "/memory/{}",
        js_sys::encode_uri_component(id),
    );
    api_delete(&path).await
}

// ── Autonomy API ────────────────────────────────────────────────────

/// Fetch autonomy settings.
pub async fn fetch_autonomy_settings() -> Result<AutonomySettings, ApiError> {
    api_fetch("/autonomy").await
}

/// Update autonomy mode.
pub async fn update_autonomy_settings(mode: &str) -> Result<AutonomySettings, ApiError> {
    #[derive(Serialize)]
    struct Body<'a> {
        mode: &'a str,
    }
    api_post("/autonomy", &Body { mode }).await
}
